File: library-management/typescript/tests/LibraryBuilder.ts
import { Library } from '../src/Library.js';
import { Member } from '../src/Member.js';
import { BookBuilder } from './BookBuilder.js';
import { FixedClock } from './FixedClock.js';
import { RecordingNotifier } from './RecordingNotifier.js';

export class LibraryBuilder {
  private _openedOn: Date = new Date(Date.UTC(2026, 0, 1));
  private _members: Member[] = [];
  private _books: BookBuilder[] = [];

  openedOn(d: Date): this { this._openedOn = d; return this; }
  withMember(m: Member): this { this._members.push(m); return this; }
  withBook(b: BookBuilder): this { this._books.push(b); return this; }

  build(): { library: Library; notifier: RecordingNotifier; clock: FixedClock } {
    const clock = new FixedClock(this._openedOn);
    const notifier = new RecordingNotifier();
    const library = new Library(clock, notifier);
    for (const m of this._members) library.seedMember(m);
    for (const b of this._books) library.seedBook(b.build(), b.copyCount);
    return { library, notifier, clock };
  }
}

File: library-management/python/tests/library_builder.py
from __future__ import annotations

from datetime import date

from library_management import Library, Member

from .book_builder import BookBuilder
from .fixed_clock import FixedClock
from .recording_notifier import RecordingNotifier


class LibraryBuilder:
    def __init__(self) -> None:
        self._opened_on: date = date(2026, 1, 1)
        self._members: list[Member] = []
        self._books: list[BookBuilder] = []

    def opened_on(self, d: date) -> "LibraryBuilder":
        self._opened_on = d
        return self

    def with_member(self, member: Member) -> "LibraryBuilder":
        self._members.append(member)
        return self

    def with_book(self, builder: BookBuilder) -> "LibraryBuilder":
        self._books.append(builder)
        return self

    def build(self) -> tuple[Library, RecordingNotifier, FixedClock]:
        clock = FixedClock(self._opened_on)
        notifier = RecordingNotifier()
        library = Library(clock, notifier)
        for m in self._members:
            library.seed_member(m)
        for b in self._books:
            library.seed_book(b.build(), b.copy_count)
        return library, notifier, clock
